### PR TITLE
Solve permission issues when performing btDiscoverAndGetResults

### PR DESCRIPTION
**Pull request by stplaydog**
This change is [Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/169)

### CODE REVIEW

File: src/main/java/com/google/android/mobly/snippet/bundled/utils/Utils.java [hunk 1]
```diff
@@ -210,4 +215,23 @@ public static String bytesToHexString(byte[] bytes) {
         }
         return new String(hexChars);
     }
+
+   public static void adaptShellPermissionIfRequired(Context context) throws Throwable {
+      if (context.getApplicationContext().getApplicationInfo().targetSdkVersion >= 29
+          && Build.VERSION.SDK_INT >= 29) {
+        Log.d("Elevating permission require to enable support for wifi operation in Android Q+");
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/utils/Utils.java:222]: why does this say wifi when you are actually using it for bt?

is this copied from somewhere?

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 2]
```diff
@@ -190,6 +191,13 @@ public String btGetName() {
         return mBluetoothAdapter.getName();
     }
 
+    @Rpc(description = "Automatically confirm the bt pairing request.")
+    public void btConfirmPair() throws Throwable {
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:195]: shouldn't this be part of the Rpc that initiates the pairing instead?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/android/mobly/snippet/bundled/utils/Utils.java (hunk 1 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/utils/Utils.java:222]: Yes, it's copied from WifiManagerSnippet, let me consolidate this function.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 2 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:195]: This is what missing from the device being paired, actually it needs a listener to autoconfirm the paring request. SL4A has a similar implementation: bluetoothStartPairingHelper

Comment (xpconanfan) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:195]: Whichever Rpc initiates the pairing should include the confirmation right?

Or is this confirming an incoming pair request? If so, the name and documentation should be clear on that.

Right now this is pretty confusing.

Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:195]: Thanks, I've changed the name and documentation to clarify the purpose of this function.

**Review by xpconanfan — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/169)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @stplaydog)

___
*[`src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java` line 195 at r1](https://reviewable.io/reviews/google/mobly-bundled-snippets/169#gh-1228496543:-NZ2Whyt4gBWIO8n6gcj:b6g5ler) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/f8e153ede200336c20d519b6746ddfadc510c9e5/src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java#L195)):*
<details><summary><i>Previously, stplaydog (Zhaoming Yin) wrote…</i></summary><blockquote>

Thanks, I've changed the name and documentation to clarify the purpose of this function.
</blockquote></details>

Something like `btStartAutoAcceptIncomingPairRequst` ?
And we should probably have a "Stop" for it as well?

Also, should the stop be part of the `shutdown`, or will the receiver automatically go away?

**Comment by stplaydog** [comment]
> > _[Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/169)_ status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @stplaydog)
> 
> _[`src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java` line 195 at r1](https://reviewable.io/reviews/google/mobly-bundled-snippets/169#gh-1228496543:-NZ2Whyt4gBWIO8n6gcj:b6g5ler) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/f8e153ede200336c20d519b6746ddfadc510c9e5/src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java#L195)):_
> 
> Previously, stplaydog (Zhaoming Yin) wrote…
> Something like `btStartAutoAcceptIncomingPairRequst` ? And we should probably have a "Stop" for it as well?
> 
> Also, should the stop be part of the `shutdown`, or will the receiver automatically go away?

roger, implement the function similar to how it's handled in SL4A:

1. changed the name
2. add top function
3. add stop in the shutdown part

* test auto confirm (SUCCESS)

target `s.btEnable() -> s.btBecomeDiscoverable(1500) -> s.btSetName("target") -> s.btStartAutoAcceptIncomingPairRequest()`

source `s.btEnable() -> s.btDiscoverAndGetResults() -> s.btPairDevice('08:9E:08:35:01:C6')`

* test cancel auto confirm (NOT SUCCESS)

target `s.btEnable() -> s.btBecomeDiscoverable(1500) -> s.btSetName("target") -> s.btStartAutoAcceptIncomingPairRequest() -> s.btStopAutoAcceptIncomingPairRequest()`

source `s.btEnable() -> s.btDiscoverAndGetResults() -> s.btPairDevice('08:9E:08:35:01:C6')`

File: src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java [hunk 3]
```diff
@@ -403,33 +406,6 @@ public void wifiDisableSoftAp() throws Throwable {
...
-     * below Android Q privacy changes for additional details.
-     * https://developer.android.com/preview/privacy/camera-connectivity
-     *
-     * @throws Throwable if failed to cleanup connection with UiAutomation
-     */
-    private void adaptShellPermissionIfRequired() throws Throwable {
-        if (mContext.getApplicationContext().getApplicationInfo().targetSdkVersion >= 29
-            && Build.VERSION.SDK_INT >= 29) {
-          Log.d("Elevating permission require to enable support for wifi operation in Android Q+");
-          UiAutomation uia = InstrumentationRegistry.getInstrumentation().getUiAutomation();
```
Comment (ko1in1u) [inline src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java:419]: Please remove Unused imports: android.app.UiAutomation

File: src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java [hunk 4]
```diff
@@ -403,33 +406,6 @@ public void wifiDisableSoftAp() throws Throwable {
...
-          uia.adoptShellPermissionIdentity();
-          try {
-            Class<?> cls = Class.forName("android.app.UiAutomation");
-            Method destroyMethod = cls.getDeclaredMethod("destroy");
-            destroyMethod.invoke(uia);
-          } catch (NoSuchMethodException
-              | IllegalAccessException
-              | ClassNotFoundException
-              | InvocationTargetException e) {
-                  throw new WifiManagerSnippetException("Failed to cleaup Ui Automation", e);
```
Comment (ko1in1u) [inline src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java:429]: Please remove Unused class: WifiManagerSnippetException

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 5]
```diff
@@ -36,7 +36,10 @@
 import com.google.android.mobly.snippet.rpc.Rpc;
 import com.google.android.mobly.snippet.util.Log;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
```
Comment (ko1in1u) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:42]: In alphabetical order.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 6]
```diff
@@ -369,7 +390,12 @@ public void btUnpairDevice(String deviceAddress) throws Throwable {
     }
 
     @Override
-    public void shutdown() {}
+    public void shutdown() {
+        for (Map.Entry<String,BroadcastReceiver> entry : mReceivers.entrySet()) {
```
Comment (ko1in1u) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:394]: Need a space between comma and BroadcastReceiver

File: src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java (hunk 4 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java:429]: WifiManagerSnippetException is still widely used in this class

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 5 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:42]: done

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 6 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:394]: done, thanks!

File: src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java (hunk 3 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java:419]: done

File: src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java (hunk 4 above)
Comment (stplaydog) [reply src/main/java/com/google/android/mobly/snippet/bundled/WifiManagerSnippet.java:429]: removed unused WifiManagerSnippetException Ctor.

**Review by ko1in1u — APPROVED** [review]

**Pull request merged**